### PR TITLE
Altar

### PR DESCRIPTION
**Pull request by Chocimier**
Hello. Here is fixed altar's bug, in simplest way, without refactoring.

### CODE REVIEW

File: CCallback.cpp [hunk 1]
```diff
@@ -218,9 +218,21 @@ void CCallback::trade(const CGObjectInstance *market, EMarketMode::EMarketMode m
 	pack.market = market;
 	pack.hero = hero;
 	pack.mode = mode;
-	pack.r1 = id1;
-	pack.r2 = id2;
-	pack.val = val1;
+	pack.r1 = std::vector<ui32>{static_cast<ui32>(id1)};
```
Comment (alexvins) [inline CCallback.cpp]: you should change method parameter type instead of casting

File: CCallback.cpp [hunk 2]
```diff
@@ -218,9 +218,21 @@ void CCallback::trade(const CGObjectInstance *market, EMarketMode::EMarketMode m
...
-	pack.r1 = id1;
-	pack.r2 = id2;
-	pack.val = val1;
+	pack.r1 = std::vector<ui32>{static_cast<ui32>(id1)};
+	pack.r2 = std::vector<ui32>{static_cast<ui32>(id2)};
+	pack.val = std::vector<ui32>{static_cast<ui32>(val1)};
+	sendRequest(&pack);
+}
+
+void CCallback::trade(const CGObjectInstance *market, EMarketMode::EMarketMode mode, std::vector<int> id1, std::vector<int> id2, std::vector<int> val1, const CGHeroInstance *hero)
```
Comment (alexvins) [inline CCallback.cpp]: Also change vector parameters to constant reference

**Comment by Chocimier** [comment]
What are those formatting issues to fix?

File: CCallback.cpp [hunk 3]
```diff
@@ -212,7 +212,19 @@ void CCallback::buyArtifact(const CGHeroInstance *hero, ArtifactID aid)
 	sendRequest(&pack);
 }
 
-void CCallback::trade(const CGObjectInstance *market, EMarketMode::EMarketMode mode, int id1, int id2, int val1, const CGHeroInstance *hero)
+void CCallback::trade(const CGObjectInstance *market, EMarketMode::EMarketMode mode, ui32 id1, ui32 id2, ui32 val1, const CGHeroInstance *hero)
```
Comment (alexvins) [inline CCallback.cpp]: add space around `*`

File: CCallback.h [hunk 4]
```diff
@@ -56,7 +56,8 @@ class IGameActionCallback
 	virtual bool upgradeCreature(const CArmedInstance *obj, SlotID stackPos, CreatureID newID=CreatureID::NONE)=0; //if newID==-1 then best possible upgrade will be made
 	virtual void swapGarrisonHero(const CGTownInstance *town)=0;
 
-	virtual void trade(const CGObjectInstance *market, EMarketMode::EMarketMode mode, int id1, int id2, int val1, const CGHeroInstance *hero = nullptr)=0; //mode==0: sell val1 units of id1 resource for id2 resiurce
+	virtual void trade(const CGObjectInstance *market, EMarketMode::EMarketMode mode, ui32 id1, ui32 id2, ui32 val1, const CGHeroInstance *hero = nullptr)=0; //mode==0: sell val1 units of id1 resource for id2 resiurce
+	virtual void trade(const CGObjectInstance *market, EMarketMode::EMarketMode mode, const std::vector<ui32> &id1, const std::vector<ui32> &id2, const std::vector<ui32> &val1, const CGHeroInstance *hero = nullptr)=0;
```
Comment (alexvins) [inline CCallback.h]: add space around `*` and `&`

File: client/windows/CTradeWindow.cpp [hunk 5]
```diff
@@ -1231,13 +1231,20 @@ void CAltarWindow::makeDeal()
 		blockTrade();
 		slider->moveTo(0);
 
-		std::vector<int> toSacrifice = sacrificedUnits;
-		for (int i = 0; i < toSacrifice.size(); i++)
+		std::vector<ui32> ids;
+		std::vector<ui32> toSacrifice;
+
+		for (int i = 0; i < sacrificedUnits.size(); i++)
```
Comment (alexvins) [inline client/windows/CTradeWindow.cpp]: remove space after `for`

File: server/CGameHandler.cpp [hunk 6]
```diff
@@ -5418,50 +5418,89 @@ void CGameHandler::visitObjectOnTile(const TerrainTile &t, const CGHeroInstance
...
 		COMPLAIN_RET("You need hero to sacrifice creature!");
 
-	int oldCount = hero->getStackCount(slot);
+	int expSum = 0;
+	auto finish = [this, &hero, &expSum]()
+	{
+		changePrimSkill(hero, PrimarySkill::EXPERIENCE, hero->calculateXp(expSum));
+	};
+
+	for (int i = 0; i < slot.size(); ++i)
```
Comment (alexvins) [inline server/CGameHandler.cpp]: remove space after `for`

File: server/CGameHandler.cpp [hunk 7]
```diff
@@ -5418,50 +5418,89 @@ void CGameHandler::visitObjectOnTile(const TerrainTile &t, const CGHeroInstance
...
+	{
+		int oldCount = hero->getStackCount(slot[i]);
+
+		if (oldCount < count[i])
```
Comment (alexvins) [inline server/CGameHandler.cpp]: remove space after `if`

**Pull request merged**